osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/test/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/utils/DlmsHelperTest.java

```diff
@@ -80,36 +83,99 @@ public class DlmsHelperTest {
     when(connectionManager.getConnection()).thenReturn(dlmsConnection);
 
-    final AttributeAddress[] attrAddresses = new AttributeAddress[1];
-    attrAddresses[0] = mock(AttributeAddress.class);
+    // Add one more address to the request than the maximum
+    final AttributeAddress[] attrAddresses = new AttributeAddress[getWithListMax + 1];
+    for (int i = 0; i <= getWithListMax; i++) {
+      attrAddresses[i] = mock(AttributeAddress.class);
+    }
+
+    this.dlmsHelper.getWithList(connectionManager, dlmsDevice, attrAddresses);
+
+    // We expect 2 calls:
+    // - one call with all addresses up to the maximum amount
+    // - one call with the remaining address
+    verify(dlmsConnection).get(Arrays.asList(attrAddresses).subList(0, getWithListMax));
+    verify(dlmsConnection).get(Collections.singletonList(attrAddresses[getWithListMax]));
+  }
 
-    when(dlmsDevice.isWithListSupported()).thenReturn(true);
+  @ParameterizedTest
+  @CsvSource({"2", "5", "32"})
+  void testGetWithListWithMoreThanTwiceTheMaximum(final int getWithListMax)
+      throws ProtocolAdapterException, IOException {
+    final DlmsConnection dlmsConnection = mock(DlmsConnection.class);
+    final DlmsConnectionManager connectionManager = mock(DlmsConnectionManager.class);
+    final DlmsDevice dlmsDevice = mock(DlmsDevice.class);
+    when(dlmsDevice.getWithListMax()).thenReturn(getWithListMax);
+    when(connectionManager.getConnection()).thenReturn(dlmsConnection);
+
+    // Add one more address to the request than twice the maximum
+    final AttributeAddress[] attrAddresses = new AttributeAddress[(getWithListMax * 2) + 1];
+    for (int i = 0; i <= (getWithListMax * 2); i++) {
+      attrAddresses[i] = mock(AttributeAddress.class);
+    }
 
     this.dlmsHelper.getWithList(connectionManager, dlmsDevice, attrAddresses);
-    verify(dlmsConnection).get(Arrays.asList(attrAddresses));
+
+    // We expect 3 calls:
+    // - one call with all addresses up to the maximum amount
+    // - another call with the maximum amount of addresses
+    // - one call with the remaining address
+    verify(dlmsConnection).get(Arrays.asList(attrAddresses).subList(0, getWithListMax));
+    verify(dlmsConnection)
+        .get(Arrays.asList(attrAddresses).subList(getWithListMax, getWithListMax * 2));
+    verify(dlmsConnection).get(Collections.singletonList(attrAddresses[getWithListMax * 2]));
   }
 
-  @Test
-  public void testGetWithListWorkaround() throws ProtocolAdapterException, IOException {
+  @ParameterizedTest
+  @CsvSource({"0", "1"})
+  void testNormalGetWithMultipleAddresses(final int getWithListMax)
+      throws ProtocolAdapterException, IOException {
     final DlmsConnection dlmsConnection = mock(DlmsConnection.class);
     final DlmsConnectionManager connectionManager = mock(DlmsConnectionManager.class);
     final DlmsDevice dlmsDevice = mock(DlmsDevice.class);
+
+    // Set the maximum to 0 or 1. When the max is 0, then it will be handled as a max of 1.
```

Comment:
Is this a test for some hidden feature? or do we miss setting this max value somewhere?

Reply:
Okay, this makes it fool proof